File: src/main/webapp/tutorials/code/typescript/generics-constraints.ts
// Generic Constraints

// Constraint with 'extends'
interface HasLength {
    length: number;
}

function logLength<T extends HasLength>(item: T): void {
    console.log("Length:", item.length);
}

logLength("Hello");  // ✓ string has length
logLength([1, 2, 3]);  // ✓ array has length
// logLength(123);  // ✗ Error - number doesn't have length

// Constraint with keyof
function getProperty<T, K extends keyof T>(obj: T, key: K): T[K] {
    return obj[key];
}

let person = { name: "Alice", age: 25 };
let personName = getProperty(person, "name");  // Type: string
let personAge = getProperty(person, "age");  // Type: number

console.log(personName, personAge);

File: src/main/webapp/tutorials/code/go/generics-constraints.go
package main

import (
	"fmt"

	"golang.org/x/exp/constraints"
)

// Custom constraint using interface
type Number interface {
	int | int64 | float64
}

// Function using custom constraint
func Sum[T Number](numbers []T) T {
	var total T
	for _, n := range numbers {
		total += n
	}
	return total
}

// Using built-in constraints
func Min[T constraints.Ordered](a, b T) T {
	if a < b {
		return a
	}
	return b
}

// Constraint with methods
type Stringer interface {
	String() string
}

func PrintAll[T Stringer](items []T) {
	for _, item := range items {
		fmt.Println(item.String())
	}
}

// Custom type implementing Stringer
type Person struct {
	Name string
	Age  int
}

func (p Person) String() string {
	return fmt.Sprintf("%s (%d years)", p.Name, p.Age)
}

func main() {
	// Using Number constraint
	ints := []int{1, 2, 3, 4, 5}
	fmt.Println("Sum of ints:", Sum(ints))

	floats := []float64{1.1, 2.2, 3.3}
	fmt.Println("Sum of floats:", Sum(floats))

	// Using Ordered constraint
	fmt.Println("\nMin examples:")
	fmt.Println("Min(10, 20):", Min(10, 20))
	fmt.Println("Min(\"apple\", \"banana\"):", Min("apple", "banana"))

	// Using interface constraint
	people := []Person{
		{"Alice", 25},
		{"Bob", 30},
		{"Charlie", 35},
	}

	fmt.Println("\nPeople:")
	PrintAll(people)
}
